### ISSUE TITLE
CanSupportCollectionInitializer

### ISSUE THREAD

**Issue by mkrueger** [dotnet/roslyn#1564]
In src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs

Shouldn't the CanSupportCollectionInitializer method check also the base types for an Add method ?

I've a similiar case and I use:

```
        public static bool CanSupportCollectionInitializer(this ITypeSymbol typeSymbol)
        {
            if (typeSymbol.AllInterfaces.Any (i => i.SpecialType == SpecialType.System_Collections_IEnumerable)) {
                var curType = typeSymbol;
                while (curType != null) {
                    if (HasAddMethod (curType))
                        return true;
                    curType = curType.BaseType;
                }
            }
            return false;
        }

        static bool HasAddMethod (ITypeSymbol typeSymbol)
        {
            return typeSymbol
                .GetMembers (WellKnownMemberNames.CollectionInitializerAddMethodName)
                .OfType<IMethodSymbol> ().Any (m => m.Parameters.Any ());
        }
```

**Comment by Pilchie** [comment]
Yes, you are probably right.  However, note that you should probably use `LookupSymbols` instead of walking up the base types.  That should handle inaccessible members, etc properly unlike your current implementation.  Care to submit a PR?

**Comment by Pilchie** [comment]
Actually, I just learned something. Whether a type have a collection initializer depends on *where* it is being used.  If the Add method is private/protected, then it can be collection initialized where Add is accessible and not elsewhere.  This means that the method probably needs a context location to ask about.

(Another thing about LookupSymbols is that it will handle extension methods - also assuming you give it a context location).

**Comment by Pilchie** [comment]
Anyway, AFAIK, the consequences of not fixing this are that we will sometimes show Object Initializer completion in a place where a collection initializer is valid, which isn't the end of the world.

**Comment by CyrusNajmabadi** [comment]
Looks like we do check the base types: https://github.com/dotnet/roslyn/blob/9592fdad9227dbc65d2add06c149788e03711811/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs#L407

**Comment by sharwell** [comment]
Fixed by #5536
